File: Dec2020/10Dec/TrimABinarySearchTree.java
/*
Problem Name: 669. Trim a Binary Search Tree
Problem Link: https://leetcode.com/problems/trim-a-binary-search-tree/
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode trimBST(TreeNode root, int low, int high) {
        return traverse(root, low, high);
    }
    TreeNode traverse(TreeNode root, int l, int h){
        if(root == null)
            return root;
        if(root.val > h)
            return traverse(root.left, l, h);
        if(root.val < l)
            return traverse(root.right, l, h);
        root.left = traverse(root.left, l, h);
        root.right = traverse(root.right, l, h);
        return root;  
    }
}

File: Dec2020/10Dec/TrimABinarySearchTree.cpp
/*
Problem name: 669. Trim a Binary Search Tree
Problem link: https://leetcode.com/problems/trim-a-binary-search-tree/
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* trimBST(TreeNode* root, int low, int high) {
        if(!root)
            return nullptr;
       
        root->left = trimBST(root->left, low, high);
        root->right = trimBST(root->right, low, high);
        if(root->val < low){
            return root->right;
        }
        if(root->val > high){
            return root->left;
        }
        return root;            
    }
};
